tools/pybind11Tools.cmake

```diff
@@ -37,25 +37,31 @@ function(select_cxx_standard)
 endfunction()
 
 # Internal: find the appropriate LTO flag for this compiler
-macro(_pybind11_find_lto_flag output_var prefer_thin_lto)
-  if(${prefer_thin_lto})
+function(_pybind11_find_lto_flag output_var prefer_thin_lto)
+  set(try_flags)
+  if (NOT CMAKE_CXX_COMPILER_ID MATCHES "Intel")
     # Check for ThinLTO support (Clang)
-    check_cxx_compiler_flag("-flto=thin" HAS_THIN_LTO_FLAG)
-    set(${output_var} $<${HAS_THIN_LTO_FLAG}:-flto=thin>)
+    if(${prefer_thin_lto})
+      list(APPEND try_flags "-flto=thin")
+    endif()
+    # Check for Link Time Optimization support (GCC/Clang)
+    list(APPEND try_flags "-flto")
+  else()
+    # Intel equivalent to LTO is called IPO
+    list(APPEND try_flags "-ipo")
   endif()
 
-  if(NOT ${prefer_thin_lto} OR NOT HAS_THIN_LTO_FLAG)
-    if(NOT CMAKE_CXX_COMPILER_ID MATCHES "Intel")
-      # Check for Link Time Optimization support (GCC/Clang)
-      check_cxx_compiler_flag("-flto" HAS_LTO_FLAG)
-      set(${output_var} $<${HAS_LTO_FLAG}:-flto>)
-    else()
-      # Intel equivalent to LTO is called IPO
-      check_cxx_compiler_flag("-ipo" HAS_IPO_FLAG)
-      set(${output_var} $<${HAS_IPO_FLAG}:-ipo>)
+  set(orig_CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}")
+  foreach(f ${try_flags})
+    # clang, in particular, needs -flto to be provided at link time as well; it shouldn't hurt the others
+    set(CMAKE_REQUIRED_LIBRARIES "${orig_CMAKE_REQUIRED_LIBRARIES} ${f}")
```

Comment:
I changed it from a macro to a function--the main difference being that cmake does an implicit push/pop of variables, so the sets here are local (the exception being the PARENT_SCOPE one later on).